### PR TITLE
[IN PROGRESS] Trusted Proxies Support

### PR DESCRIPTION
**Pull request by msjyoo**
Hello,

I am working on implementing trusted proxies support, for reverse proxies and X-Forwarded-For headers.

Putting this quick prototype here so others comment on the implementation before I work too much on it.

TODO:
- [x] Write Unit Tests for the code written
  - [x] Write unit test for Instantiation
  - [x] Write unit test for IPv4 validation
  - [x] Write unit test for IPv6 validation
  - [x] Modify unit tests for other classes
- [ ] Test in production
- [x] Add TrustedProxies Interface
- [ ] NEW: Make sure that the API is as easy to work with as possible

Removed from TODO:
- Implement #892
- Implement HTTP Forwarded header: http://tools.ietf.org/html/rfc7239

Please note that I will be squashing all the commits before the final request for merge.

### CODE REVIEW

**Comment by DavidePastore** [comment]
Nice to see this. It is associated with #892

**Comment by codeguy** [comment]
Any progress on this?

**Comment by msjyoo** [comment]
@DavidePastore I think #892 is unnecessary on the new 3.0.0 since the getHost() uses the URI itself not the HTTP_HOST environment variable.

**Comment by msjyoo** [comment]
It seems that no reasonable proxy actually uses the Forwarded header, so I am leaving that off the todo list for now.

**Comment by DavidePastore** [comment]
@sekjun9878 In my case **HTTP_X_FORWARDED_HOST** was used from the reverse proxy server (Apache).

**Comment by msjyoo** [comment]
@DavidePastore Yes, but is the HTTP_X_FORWARDED_HOST different from the URL host provided in the HTTP request URL?

**Comment by DavidePastore** [comment]
@sekjun9878 Not simple to remember. It was a long time ago when I faced this.

**Comment by codeguy** [comment]
How's this coming?

**Comment by msjyoo** [comment]
I'm still working on it, I've just had other things to do on my hand.. If you want to take over from here then feel free to checkout what I have and expand on it. If not, then I'll hopefully finish it in about a week or two and make a final PR.

**Comment by codeguy** [comment]
Week or two sounds great! Just checking in. Will probably roll out a beta release earlier though and merge this in closer to release.

File: tests/AppTest.php [hunk 1]
```diff
@@ -499,6 +500,26 @@ public function testInvokeWithPimpleUndefinedCallable()
         $app($req, $res);
     }
 
+	/********************************************************************************
+	 * Client Details / Reverse Proxy Handling
+	 *******************************************************************************/
+
+	public function testGetClientIp()
```
Comment (msjyoo) [inline tests/AppTest.php:507]: Edit: Nevermind, didn't notice that there were new composer dependencies added.

Comment (codeguy) [inline tests/AppTest.php:507]: All tests are green on my end. Can you post the error output you're seeing on your end?

**Comment by codeguy** [comment]
Any updates? Just fyi I'll be tagging 3.0.0-beta1 in a day or two.

**Comment by msjyoo** [comment]
@codeguy Unfortunately I don't have time to finish and test this PR - it's nearly finished anyway, so if you would like to just cleanup the code and test / merge then please do. Sorry!

**Pull request closed without merging**